src/grain/ledger.js

```diff
@@ -0,0 +1,146 @@
+// @flow
+
+import {type NodeAddressT, NodeAddress} from "../core/graph";
+import stringify from "json-stable-stringify";
+import * as NullUtil from "../util/null";
+import {
+  type Grain,
+  format as formatGrain,
+  ZERO,
+  DECIMAL_PRECISION,
+} from "./grain";
+import {type DistributionV1} from "./distribution";
+
+/**
+ * Models a transfer of grain from a sender to recipient.
+ * A grainholder may transfer grain to themself (which is a no-op).
+ */
+export type TransferV1 = {|
+  +type: "TRANSFER",
+  +version: number,
+  +sender: NodeAddressT,
+  +recipient: NodeAddressT,
+  +amount: Grain,
+  +timestampMs: number,
+|};
+
+export type LedgerEvent = DistributionV1 | TransferV1;
```

Comment:
I'm wondering if "Distribution" actually belongs as a ledger event vs having only the lower level functions of "Mint" and "Transfer", which is a lot more generic and composable. The "Distribution" would essentially be a set of "Mint" events for specific node addresses. Not sure if the ledger needs to / should "know" about distributions, it seems like the only thing that would matter would be the list of "Receipts" aka minting grain.

Conceptually as well, "Distribution" is not a "ledger event" as much as it is a "collection of ledger events", vs its very obvious that a "Transfer" is a ledger event.

Reply:
I thought about this. The issue is for the UI that we're going to build this for, we're going to want to be able to find all the distributions for a user that gave them grain. So if we don't keep that info in the ledger, then we need to track two ledgers: the grain balance ledger, and the distribution history ledger. At which point we risk them de-synchronizing, so we might want to bind them more tightly by giving every LedgerEvent an ID and referencing the distribution IDs from the distribution history. But adding IDs adds even more complexity.

Given that this is a "private" data format to SourceCred, and that that the simpler ledger is derivable from a transformation on this ledger version, we also could always upgrade into that format later (and just provide a migration script). 

Thus, I'm inclined to keep it as is. Happy to hear alternative perspectives, though.

Reply:
Good question @hammadj. And I'm inclined to agree with @decentralion, but with a slightly different rationale.

---

First, after having another look. I want to get some terminology straight.

The `Ledger` interface and `InMemoryLedger` class. **Are not ledgers**. The `Ledger` interface is actually a _projection_, the result of aggregating a stream of events into a "current state".

It includes data a ledger does not have: reduced state. And it's missing the only, but required mutation a ledger has: append.

I think a ledger is an append-only log of _account mutations_. This is a specific variant of the append-only log of events. The term borrows from the (much more mature than IT) field of accounting.

---

> So if we don't keep that info in the ledger, then we need to track two ledgers: the grain balance ledger, and the distribution history ledger. At which point we risk them de-synchronizing,

This is very much on point. You can definitely create working systems with multiple event logs. I've built a couple of these in the past. But the main concern when deciding whether it's best to have one or multiple is _consistency guarantees_.

Consistency in an event log all boils down to _the order of events_.

Using one log is simple, because a log is ordered. The sequence number / array index is what gives you consistency.

When you're using multiple logs, the only approximation of order you have is the timestamp. Timestamps open you up to all kinds of race conditions and edge cases. Referencing IDs does not solve this. And things get very messy when considering a distributed system.

That said, I think @hammadj's suggested approach is feasible, but comes at a cost. Representing a distribution as a sequence of "Grain Minted" events would work. Without the second event log, this is a lossy representation.

You can _reconstruct_ the distribution by projecting this sequence of minting events back into what the actual effect was of this sequence. And you can answer the question of whether these events were part of the same distribution using reference IDs. But what can't be reconstructed is whether the actual effect matches the intended effect. So preserving this intent is what we'd need the second event log for.

By modeling the distribution as a single event, you don't lose the context or intent.

I think we _want to_ provide consistency guarantees about the distribution as a whole and it's intent. Using a single event and one log, is actually a less complex way to achieve that than requiring the second log and dealing with inconsistency.

---

As an aside. Computational, implementation or event model complexity are not very important. Taking time to carefully model your events is definitely worthwhile though, because changing events after the fact is illegal.

For better or worse you will be supporting these event models _until the end of time_. I disagree with @decentralion here, with regards to "private data" and upgrade paths. These events are your _source of truth_ and updating them after the fact is a breach of contract. Even re-encoding the same data with the same meaning into a new format should be frowned upon. If this were on-chain data, this would rightfully be made infeasible to do. The only thing you can upgrade is how you interpret this data, and fix mistakes by appending corrections.

If this "simpler ledger" you had in mind was actually not a ledger but a projection though. That's something you can do. That's an upgrade of how you interpret the data. Discarding data and making use case specific projections is totally OK (and a feature in many cases).

Finding those caching and optimization strategies is something you can do later and as needed. But reproducing data that was lost because the model was incomplete when you recorded it is something you want to avoid.

So as a rule of thumb, I would side with caution and include data you can't easily reproduce and seems like it would have a future use case.

For example, something I like about the current model: it includes the starting budget and (versioned reference to) the strategy used. But doesn't include easy to reproduce data such as the sum of actually minted grain and the difference from the starting budget.
But I'll get into that in other comments.